R/chat.R

```diff
@@ -487,15 +487,22 @@ Chat <- R6::R6Class(
       stream,
       echo
     ) {
+      tool_errors <- list()
+      withr::defer(warn_tool_errors(tool_errors))
+
       while (!is.null(user_turn)) {
-        for (chunk in private$submit_turns(
-          user_turn,
-          stream = stream,
-          echo = echo
-        )) {
+        # fmt: skip
+        for (chunk in private$submit_turns(user_turn, stream = stream, echo = echo)) {
           yield(chunk)
         }
-        user_turn <- private$invoke_tools()
+
+        user_turn <- private$invoke_tools(echo = echo)
+
+        if (echo == "all") {
+          cat(format(user_turn))
+        } else if (echo == "none") {
+          tool_errors <- c(tool_errors, turn_get_tool_errors(user_turn))
+        }
```

Comment:
Initially, I handled echoing tool results at this level, but we would end up showing all requests and then collectively showing all results. Passing `echo` into `invoke_tools()` lets us move the display closer to where it happens so that request/results are paired together.

On the other hand, I thought it'd be better to collect all tool errors in the entire round and then emit one warning with a list of errors at the end of the round.